scripts/build.js

```diff
@@ -116,9 +115,28 @@ fs.mkdirSync(outdir, { recursive: true });
         routes: {
           '/dist': './dist'
         }
+      },
+      // Configure socketIO to retry forever when disconnected to enable the auto-reattach timeout below to work
+      socket: {
+        socketIoClientConfig: {
+          reconnectionAttempts: Infinity,
+          reconnectionDelay: 500,
+          reconnectionDelayMax: 500,
+          timeout: 1000
+        }
       }
     });
 
+    setTimeout(() => {
+      const url = `http://localhost:${port}`;
+      console.log(chalk.cyan(`Launched the Shoelace dev server at ${url} 🥾\n`));
+      if (Object.keys(bs.sockets.sockets).length === 0) {
+        open(url);
+      } else {
+        bs.reload();
+      }
+    }, 2000);
```

Comment:
So here's basically how the process works in my mind. Each open client has a websocket connection to the dev server. When the server shuts down, it is configured above to retry infinitely with a 500ms delay and a 1,000ms timeout. So within 2,000ms, every open tab should have an opportunity to retry at least once and reconnect. So the napkin math I used was:

```
const retryDelay = 500;
const retryTimeout = 1000;
const openNewTabTimeout = retryDelay * 2 + retryTimeout;
```

It's sort of arbitrary, but also not completely. One thing which this doesn't account for is the time it takes for the dev server to start because `bs.init` is async. If we moved this code inside the second argument of `.init` (the done callback) then it should account for that time too and then if the build takes over 2s for some reason, it should hopefully still work. But I wasn't able to see any difference when putting the code in the callback since the build is just so dang fast currently haha.

I didn't reference any other library's implementations for this, but just theory crafting, I'm not sure how to detect detatched clients without some sort of "magic number" delay to wait for them to hopefully reconnect. If you have any other approaches that don't rely on fixed timeouts, I'm all ears! 😄

Reply:
Actually, after looking at the code again, I think the latency issue is almost non-existant. The bs instance is serving from a directory on disk which has already been compiled before this line. So all the `bs.init` needs to do is start the local server, nothing needs to be compiled so the speed of that won't change as the project does. The BS docs aren't great, but it seems like [from their source code](https://github.com/BrowserSync/browser-sync/blob/master/packages/browser-sync/lib/browser-sync.js#L262-L265) that the init callback is called after everything has been setup for the server. So I'll move this timeout into that callback, add some comments to clarify, and remove the magic numbers.